sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions.RESOLVE_DIRECTORY;
+import static org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions.RESOLVE_FILE;
+import static org.apache.beam.sdk.util.MimeTypes.BINARY;
+
+import com.google.auto.value.AutoValue;
+import com.google.protobuf.ByteString;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.View;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * WholeFileIO.
+ */
+public class WholeFileIO {
+
+  public static Read read() {
+    return new AutoValue_WholeFileIO_Read.Builder().build();
+  }
+
+  public static Write write() {
+    return new AutoValue_WholeFileIO_Write.Builder().build();
+  }
+
+  /**
+   * Implements read().
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<KV<String, byte[]>>> {
+    @Nullable
+    abstract ValueProvider<String> getFilePattern();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setFilePattern(ValueProvider<String> filePattern);
+
+      abstract Read build();
+    }
+
+    public Read from(String filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return from(ValueProvider.StaticValueProvider.of(filePattern));
+    }
+
+    public Read from(ValueProvider<String> filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return toBuilder().setFilePattern(filePattern).build();
+    }
+
+    @Override
+    public PCollection<KV<String, byte[]>> expand(PBegin input) {
+      checkNotNull(
+          getFilePattern(),
+          "Need to set the filePattern of a WholeFileIO.Read transform."
+      );
+
+      String filePattern = getFilePattern().get();
+
+      PCollection<String> filePatternPCollection = input.apply(Create.of(filePattern));
+
+      PCollection<ResourceId> resourceIds = filePatternPCollection.apply(
+          ParDo.of(
```

Comment:
We have a transform for this https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/Match.java